streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java

```diff
@@ -932,14 +937,59 @@ private ProcessorTopology build(final Set<String> nodeGroup) {
                 }
             }
         }
+        processorTopology = new ProcessorTopology(new ArrayList<>(processorMap.values()),
+                topicSourceMap,
+                topicSinkMap,
+                new ArrayList<>(stateStoreMap.values()),
+                new ArrayList<>(globalStateStores.values()),
+                storeToChangelogTopic,
+                repartitionTopics,
+                stateStoreToChangelogTopicOnlyForRestoring);
+        optimizeTopology();
+        return processorTopology;
+    }
+
+    /**
+     * Topology optimizer that avoids creating changelog topics for state stores that are directly piped to a sink topic
+     */
+    private void optimizeTopology() {
+        // Find processor nodes with changelog stores
+        List<ProcessorNode> nodesWithChangeLogState = new ArrayList<>();
+        for (String key : storeToChangelogTopic.keySet()) {
+            for (ProcessorNode processorNode: processorTopology.processors()) {
+                if (processorNode.stateStores != null && processorNode.stateStores.contains(key)) {
+                    nodesWithChangeLogState.add(processorNode);
+                }
+            }
+        }
 
-        return new ProcessorTopology(new ArrayList<>(processorMap.values()),
-                                     topicSourceMap,
-                                     topicSinkMap,
-                                     new ArrayList<>(stateStoreMap.values()),
-                                     new ArrayList<>(globalStateStores.values()),
-                                     storeToChangelogTopic,
-                                     repartitionTopics);
+        // If child nodes of nodesWithChangeLogState are of SinkNode type, then we can apply the optimization.
+        for (ProcessorNode processorNode: nodesWithChangeLogState) {
+            for (ProcessorNode child: (List<ProcessorNode>) processorNode.children()) {
+                if (child instanceof SinkNode) {
+                    for (String stateStore: (Set<String>) processorNode.stateStores) {
+                        stateStoreToChangelogTopicOnlyForRestoring.put(stateStore, storeToChangelogTopic.get(stateStore));
+                    }
+                    storeToChangelogTopic.keySet().removeAll(processorNode.stateStores);
+                }
+            }
+        }
+
+        // KTable.to() uses toStream() method internally. So, we check if processorNode has child of this
+        // specific type which proceeds to Sink Node
+        for (ProcessorNode processorNode: nodesWithChangeLogState) {
+            for (ProcessorNode child: (List<ProcessorNode>) processorNode.children()) {
+                if (child.name().startsWith(KTableImpl.TOSTREAM_NAME)) {
+                    List<ProcessorNode> shouldBeSinkNode = child.children();
```

Comment:
Since shouldBeSinkNode.get(0) may not be SinkNode, consider renaming the variable